executor/aggregate.go

```diff
@@ -45,39 +106,487 @@ type HashAggExec struct {
+	}
+}
+
+func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (err error) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, w.aggCtxsMap)
+		for i, af := range w.aggFuncs {
+			if af.Update(aggEvalCtxs[i], sc, row) != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
 	return nil
 }
 
+// shuffleIntermData shuffles the intermediate data of partial workers to corresponded final workers.
+// We only support parallel execution for single-machine, so process of encode and decode can be skipped.
+func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, finalConcurrency int) {
+	groupKeysSlice := make([][][]byte, finalConcurrency)
+	for groupKey := range w.aggCtxsMap {
+		groupKeyBytes := []byte(groupKey)
+		finalWorkerIdx := int(murmur3.Sum32(groupKeyBytes)) % finalConcurrency
+		if groupKeysSlice[finalWorkerIdx] == nil {
+			groupKeysSlice[finalWorkerIdx] = make([][]byte, 0, len(w.aggCtxsMap)/finalConcurrency)
+		}
+		groupKeysSlice[finalWorkerIdx] = append(groupKeysSlice[finalWorkerIdx], groupKeyBytes)
+	}
+
+	for i := range groupKeysSlice {
+		if groupKeysSlice[i] == nil {
+			continue
+		}
+		w.outputChs[i] <- &HashAggIntermData{
+			groupKeys:   groupKeysSlice[i],
+			groupCtxMap: w.aggCtxsMap,
+		}
+	}
+}
+
+// getGroupKey evaluates the group items and args of aggregate functions.
+func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
+	w.groupValDatums = w.groupValDatums[:0]
+	for _, item := range w.groupByItems {
+		v, err := item.Eval(row)
+		if item.GetType().Tp == mysql.TypeNewDecimal {
+			v.SetLength(0)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		w.groupValDatums = append(w.groupValDatums, v)
+	}
+	var err error
+	w.groupKey, err = codec.EncodeValue(sc, w.groupKey[:0], w.groupValDatums...)
+	return w.groupKey, errors.Trace(err)
+}
+
+func (w baseHashAggWorker) getContext(sc *stmtctx.StatementContext, groupKey []byte, mapper aggCtxsMapper) []*aggregation.AggEvaluateContext {
+	aggCtxs, ok := mapper[string(groupKey)]
+	if !ok {
+		aggCtxs = make([]*aggregation.AggEvaluateContext, 0, len(w.aggFuncs))
+		for _, af := range w.aggFuncs {
+			aggCtxs = append(aggCtxs, af.CreateContext(sc))
+		}
+		mapper[string(groupKey)] = aggCtxs
+	}
+	return aggCtxs
+}
+
+func (w *HashAggFinalWorker) fetchIntermData(sc *stmtctx.StatementContext) (err error) {
+	var (
+		input *HashAggIntermData
+		ok    bool
+	)
+	for {
+		if input == nil {
+			select {
+			case <-w.finishCh:
+				return
+			case input, ok = <-w.inputCh:
+				if ok && w.intermDataRowsBuffer == nil {
+					w.intermDataRowsBuffer = make([]types.DatumRow, 0, w.maxChunkSize)
+				}
+			}
+		}
```

Comment:
1. we can totally consume this `input` before we fetch another `input`
2. it's better to extract another function to update every aggregate function on `input`